crates/assists/src/handlers/extract_assignment.rs

```diff
@@ -0,0 +1,137 @@
+use syntax::{
+    ast::{edit::AstNodeEdit, make, BinOp, BlockExpr, ElseBranch, Expr, IfExpr, NameRef, Stmt},
+    AstNode,
+};
+
+use crate::{
+    assist_context::{AssistContext, Assists},
+    AssistId, AssistKind,
+};
+
+// Assist: extract_assignment
+//
+// Extracts variable assigment to outside an if or match statement.
+//
+// ```
+// fn main() {
+//     let mut foo = 6;
+//
+//     if true {
+//         <|>foo = 5;
+//     } else {
+//         foo = 4;
+//     }
+// }
+// ```
+// ->
+// ```
+// fn main() {
+//     let mut foo = 6;
+//
+//     foo = if true {
+//         5
+//     } else {
+//         4
+//     }
+// }
+// ```
+pub(crate) fn extract_assigment(acc: &mut Assists, ctx: &AssistContext) -> Option<()> {
+    let name = ctx.find_node_at_offset::<NameRef>()?.to_string();
+
+    let if_statement = ctx.find_node_at_offset::<IfExpr>()?;
+
+    let new_stmt =
+        exprify_if(&if_statement, &name.to_string())?.indent(if_statement.indent_level());
+    let expr_stmt = make::expr_stmt(new_stmt);
+
+    acc.add(
+        AssistId("extract_assignment", AssistKind::RefactorExtract),
+        "Extract assignment",
+        if_statement.syntax().text_range(),
+        move |edit| {
+            edit.replace(if_statement.syntax().text_range(), format!("{} = {}", name, expr_stmt));
+        },
+    )
+}
+
+fn exprify_if(statement: &IfExpr, name: &str) -> Option<Expr> {
+    let then_branch = exprify_block(statement.then_branch()?, name)?;
+    let else_branch = match statement.else_branch()? {
+        ElseBranch::Block(block) => ElseBranch::Block(exprify_block(block, name)?),
+        ElseBranch::IfExpr(expr) => {
+            ElseBranch::IfExpr(IfExpr::cast(exprify_if(&expr, name)?.syntax().to_owned())?)
```

Comment:
A test for this chained `else if` part would be nice I think, could also make use of a [mark](https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/test_utils/src/mark.rs).